stl/inc/format

```diff
@@ -445,103 +449,239 @@ _NODISCARD constexpr bool _Is_execution_charset_utf8() {
 }
 
-inline constexpr bool _Is_execution_charset_utf8_v = _Is_execution_charset_utf8();
+inline constexpr char16_t _Width_estimate_low_intervals[] = { // Per N4885 [format.string.std]/11
+    0x1100u, 0x1160u, 0x2329u, 0x232Bu, 0x2E80u, 0x303Fu, 0x3040u, 0xA4D0u, 0xAC00u, 0xD7A4u, 0xF900u, 0xFB00u, 0xFE10u,
+    0xFE1Au, 0xFE30u, 0xFE70u, 0xFF00u, 0xFF61u, 0xFFE0u, 0xFFE7u};
 
-_NODISCARD constexpr int _Utf8_code_units_in_next_character(
-    const char* const _First, const char* const _Last) noexcept {
-    // Returns a count of the number of UTF-8 code units that compose the first encoded character in [_First, _Last),
-    // or -1 if [_First, _Last) doesn't contain an entire encoded character or *_First is not a valid lead byte.
-    const auto _Ch = static_cast<unsigned char>(*_First);
-    if (_Ch < 0b1000'0000u) {
-        return 1;
+inline constexpr char32_t _Width_estimate_high_intervals[] = { // Per N4885 [format.string.std]/11
+    0x1F300u, 0x1F650u, 0x1F900u, 0x1FA00u, 0x20000u, 0x2FFFEu, 0x30000u, 0x3FFFEu};
+
+template <auto& _Bounds>
+_NODISCARD constexpr int _Unicode_width_estimate(const char32_t _Ch) noexcept {
+    // Computes the width estimation for Unicode characters from N4885 [format.string.std]/11
+    int _Result = 1;
+    for (const auto& _Bound : _Bounds) {
+        if (_Ch < _Bound) {
+            return _Result;
+        }
+        _Result ^= 1;
     }
 
-    const auto _Len = static_cast<size_t>(_Last - _First);
+    return 1;
+}
+
+template <class _CharT, bool _Statically_Utf8 = _Is_execution_charset_utf8()>
+class _Fmt_codec;
+
+template <bool _Statically_Utf8>
+class _Fmt_codec_base {};
 
-    if (_Ch < 0b1110'0000u) {
-        // check for non-lead byte or partial 2-byte encoded character
-        return (_Ch >= 0b1100'0000u && _Len >= 2) ? 2 : -1;
+template <>
+class _Fmt_codec_base<false> {
+protected:
+    _Cvtvec _Cvt;
+
+    _NODISCARD int _Double_byte_encoding_code_units_in_next_character(
+        const char* const _First, const char* const _Last) const {
+        // Returns a count of the number of code units that compose the first encoded character in [_First, _Last),
+        // or -1 if [_First, _Last) doesn't contain an entire encoded character or *_First is not a valid lead byte.
+        wchar_t _Wide;
+        mbstate_t _St{};
+        const auto _Len   = static_cast<size_t>(_Last - _First);
+        const int _Result = _Mbrtowc(&_Wide, _First, _Len, &_St, &_Cvt);
+        if (_Result > 0) {
+            return _Result;
+        } else if (_Result < 0) { // invalid or incomplete encoded character
+            return -1;
+        } else { // next code unit is '\0'
+            return 1;
+        }
     }
 
-    if (_Ch < 0b1111'0000u) {
-        // check for partial 3-byte encoded character
-        return (_Len >= 3) ? 3 : -1;
+    _Fmt_codec_base() {
+        if (__std_get_cvt(__std_code_page::_Acp, &_Cvt) != __std_win_error::_Success) {
+            _THROW(format_error("FIXME"));

```

Comment:
I just want to throw a "FIXME" in here

Reply:
Yeah, this was "`format_error` is wrong here, but the Standard says I can't throw anything else, and I think it's impossible for `GetCPInfoExW` to fail when given `CP_ACP` anyhow (it's never an invalid parameter), and I'm ready to go eat dinner now."

I'm just going to ignore the return value when not testing.

Reply:
For future reference: "FIXME" is a perfectly valid review comment when there's a `FIXME` in the code.